### PR TITLE
feat(todo): enhance todo functionality and error handling

### LINKED ISSUE #197: bug(note): Recon note system is completely buggy
### Current Behavior

While I'm doing unit tests, there a lot of method and functions that do n,ot test correctly the datas.
Eg: if note is not found, nothing ...
missing parameter, nothing ....
So I've totally improved the Note section 

### Expected Behavior

Note should work and UX design well

### Steps To Reproduce

Use the note system

### Environment

```markdown
- reNgine: 2.1.0
- OS: Debian 12
- Python: 
- Docker Engine: 
- Docker Compose: 
- Browser:
```

### Anything else?

_No response_

### Acknowledgements

- [X] This issue is not a duplicate of an existing bug report.
- [X] I have chosen an appropriate title.
- [X] All requested information has been provided properly.

### PR DESCRIPTION
**Pull request by psyray**
Fixes #197 

- Refactored the JavaScript code for creating and displaying todo items to use a template-based approach.
- Improved error handling and user feedback for adding, deleting, and updating todo items.
- Updated API endpoints and views to include better validation and error messages.
- Enhanced the UI with search maintained across section, new icons and required fields for better user experience.
- Added a hidden template for todo items to streamline the creation of new items dynamically.

## Preview
image
Screenshot_20240910_102402
Screenshot_20240910_102411
Screenshot_20240910_102721
Screenshot_20240910_102733
Screenshot_20240910_102946
Screenshot_20240910_112806

## Summary by Sourcery

Enhance the todo functionality by adopting a template-based approach for item creation, improving error handling, and refining the UI with new icons and required fields. Refactor JavaScript code for better organization and introduce detailed docstrings in the views for improved documentation.

New Features:
- Introduce a template-based approach for creating and displaying todo items, enhancing the dynamic creation of new items.
- Add new UI elements, including icons and required fields, to improve user experience when interacting with todo items.

Bug Fixes:
- Improve error handling and user feedback for operations such as adding, deleting, and updating todo items, ensuring better validation and error messages.

Enhancements:
- Refactor JavaScript code to streamline event listeners and improve code organization, including the addition of functions for handling button actions and search functionality.
- Enhance the todo list UI with a hidden template for todo items, allowing for more efficient rendering and updates.

Documentation:
- Add detailed docstrings to the views in the recon_note app, providing clear explanations of the functions and their parameters.

### CODE REVIEW

File: web/recon_note/static/note/js/todo.js
[the remark this answers is not in the document]
Comment (AnonymousWP) [reply web/recon_note/static/note/js/todo.js:1]: Adding a note doesn't work for me:

image

image

Logs:

image

`make logs` doesn't show any error though.

**Review by AnonymousWP — CHANGES_REQUESTED** [review]
Tested. A general but important suggestion though: please change it to 'to-do'. 'Todo' isn't a word at all.

File: web/recon_note/static/note/js/todo.js
[the remark this answers is not in the document]
Comment (psyray) [reply web/recon_note/static/note/js/todo.js:1]: As it rely mainly on JavaScript, you need to refresh the todo.js.
So browser cache cleaning is mandatory.
Js is located at `/staticfiles/custom/todo.js`
So open the Js in your browser and do a force refresh, Ctrl+shit+r on Firefox.

[the remark this answers is not in the document]
Comment (AnonymousWP) [reply web/recon_note/static/note/js/todo.js:1]: I did, and even in incog it gives the same error.

[the remark this answers is not in the document]
Comment (psyray) [reply web/recon_note/static/note/js/todo.js:1]: ok i'll test it

[the remark this answers is not in the document]
Comment (psyray) [reply web/recon_note/static/note/js/todo.js:1]: > In my opinion, if I mark this as completed, it should be gone from the todo list and go to `Done` instead. Now it's still there at both tabs.
> 
> image

No the Todo menu displays all the todos

[the remark this answers is not in the document]
Comment (psyray) [reply web/recon_note/static/note/js/todo.js:1]: > Adding a note doesn't work for me:
> 
> image
> 

Fixed
You could pull changes and run
`docker exec -it rengine-web-1 bash -c 'poetry run -C $HOME python3 manage.py collectstatic --noinput'`
And refresh todonote.js
Or simply restart all the container

[the remark this answers is not in the document]
Comment (AnonymousWP) [reply web/recon_note/static/note/js/todo.js:1]: Fixed indeed, but now there's another thing:

It says that the subdomain is optional, but it actually isn't optional as I'm not allowed to create the todo.

image

[the remark this answers is not in the document]
Comment (psyray) [reply web/recon_note/static/note/js/todo.js:1]: Oh yes didn't see this one. I fix it

[the remark this answers is not in the document]
Comment (psyray) [reply web/recon_note/static/note/js/todo.js:1]: Fixed
All tests passed
You could continue review

[the remark this answers is not in the document]
Comment (psyray) [reply web/recon_note/static/note/js/todo.js:1]: > Fixed indeed, but now there's another thing:
> 
> It says that the subdomain is optional, but it actually isn't optional as I'm not allowed to create the todo.
> 

scan_id has also been removed from check

File: web/static/custom/custom.css
[the remark this answers is not in the document]
Comment (psyray) [reply web/static/custom/custom.css:1]: Fixed

**Comment by AnonymousWP** [comment]
It also misses two crucial things:

1. There's no 'close' button anywhere:
image
2. You can't edit notes. Pretty much a must to me, if I'm going to be honest.

**Comment by psyray** [comment]
> It also misses two crucial things:
> 
>     1. There's no 'close' button anywhere:
> 
>     2. You can't edit notes. Pretty much a must to me, if I'm going to be honest.

Done for 1.
image

2. will be done later

**Pull request merged**